server/controllers/stock/lots.js

```diff
@@ -198,12 +201,80 @@ async function merge(req, res, next) {
       transaction.addQuery(deleteLot, [mergeLotUuid]);
     });
     await transaction.execute();
-    res.sendStatus(200);
+    if (!req.params.internal) {
+      res.sendStatus(200);
+    }
   } catch (error) {
     next(error);
   }
 }
 
+/**
+ * GET /lots/merge/auto
+ *
+ * @description
+ * Finds and merges all lots suitable for automatic merging
+ *  - To qualify, the lots must have the same inventory_uuid,
+ *    label, and expiration date.
+ */
+function autoMerge(req, res, next) {
+  // The first query gets the inventory UUID for each
+  // inventory article with duplicate lots (having the
+  // same label, inventory_uuid, and expiration_date).
+  // (Since these are grouped, only one of the several
+  // lots is given.)
+  const query1 = `
+  SELECT
+    BUID(l.uuid) AS uuid, l.label, l.expiration_date,
+    BUID(i.uuid) AS inventory_uuid, i.text as inventory_text,
+    COUNT(*) as num_duplicates
+  FROM lot l
+  JOIN inventory i ON i.uuid = l.inventory_uuid
+  GROUP BY label, inventory_uuid, expiration_date HAVING num_duplicates > 1
+`;
+  // The second query gets all lots matching the label,
+  // inventory_uuid, and expiration dates
+  const query2 = `
+    SELECT
+      BUID(l.uuid) AS uuid, l.label, l.expiration_date,
+      BUID(i.uuid) AS inventory_uuid, i.text as inventory_text
+    FROM lot l
+    JOIN inventory i ON i.uuid = l.inventory_uuid
+    WHERE l.label=? AND i.uuid=? AND l.expiration_date=DATE(?)
+  `;
+  db.exec(query1, [])
+    .then((rows) => {
+      const numInventories = rows.length;
+      let numLots = rows.reduce((sum, row) => {
+        return sum + row.num_duplicates;
+      }, 0) - numInventories;
+      rows.forEach(async (row) => {
+        db.exec(query2, [row.label, db.bid(row.inventory_uuid), row.expiration_date])
+          .then((lots) => {
+            // Arbitrarily pick first lot and merge the duplicates into it
+            const masterLotUuid = lots[0].uuid;
+            const lotUuids = lots.reduce((list, elt) => {
+              if (elt.uuid !== masterLotUuid) {
+                list.push(elt.uuid);
+              }
+              return list;
+            }, []);
+            numLots = numLots + lotUuids.length - 1;
+
+            // Now merge this lot and its duplicates
+            merge({
+              query : {},
```

Comment:
You seem to be "faking" an HTTP request here (since `merge()` looks like a standard expressjs http middleware function).

I find it much simpler/more legible to simply move the "merge" logic into it's own function, then make two http interfaces that call it.

Basically, you can imagine functions:
```js
// this guy has the `db.transaction()` call and 90% of the code from the original `merge()` fn
function merge(masterLotUuid /* uuid */, duplicateLots /* array */) { /* ... */ }

// this guy wraps the above in a try/catch and breaks out the req.body/req.params 
// into the masterLotUuid and duplicateLots array
function mergeDuplicateLotsHTTP(req, res, next) { /* ... */ };

// this guy wraps the `merge()` and performs the relevant DB queries to find all duplicates for all lots.
function autoMergeDuplicateLotsHTTP(req, res, next) { /* ... */ }
```

Then it is the HTTP interfaces responsibility to:
  1) unpack the req.params/req.body to get the masterLotUuid and the duplicate lots
  2) do whatever preprocessing needed.
  3) wrap the call to `merge()` in try/catch or `.catch()` to handle errors.
  4) unpack whatever the return result is and send to appropriate status code to the client.
  
What do you think?

You can see this structure play out in our `patients` controller.  The [`find()` function](https://github.com/IMA-WorldHealth/bhima/blob/1819d8f7380f16590fce3fbbe3c900513d001c56/server/controllers/medical/patients/index.js#L742) was originally a `function (req, res, next) {}` expressjs middleware function.  But we realized we needed the same logic elsewhere, so we broke it out.  The HTTP interface for it is now called [`read()`](https://github.com/IMA-WorldHealth/bhima/blob/1819d8f7380f16590fce3fbbe3c900513d001c56/server/controllers/medical/patients/index.js#L850) and just does some basic logic switches and error handling.

Reply:
Thanks!  I applied this fix and now it works!